### PR TITLE
[WIP] Adding encryption decorator

### PR DESCRIPTION
**Pull request by prisis**
| Question      | Answer
| ------------- | ------
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #8
| License       | MIT
| Doc PR        | 

### Description
Adding a encryption decorator for psr6

### TODO
* [x] Add tests
* [ ] Add documentation
* [x] Updated Changelog.md
* [x] Rename package from encrypted to encryption

### CODE REVIEW

File: src/Encrypted/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+### Install
+
+```bash
+composer require cache/encrypted-cache
+```
+
+### Use
+
+Read the [documentation on usage](http://www.php-cache.com/en/latest/tagging/).
```
Comment (Nyholm) [inline src/Encrypted/README.md]: Update link

File: src/Encrypted/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+### Implement
+
+Read the [documentation on implementation](http://www.php-cache.com/en/latest/implementing-cache-pools/tagging/).
```
Comment (Nyholm) [inline src/Encrypted/README.md]: Update link

File: src/Encrypted/README.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+[![codecov.io](https://codecov.io/github/php-cache/encrypted-cache/coverage.svg?branch=master)](https://codecov.io/github/php-cache/encrypted-cache?branch=master)
+[![Total Downloads](https://poser.pugx.org/cache/encrypted-cache/downloads)](https://packagist.org/packages/cache/encrypted-cache)
+[![Monthly Downloads](https://poser.pugx.org/cache/encrypted-cache/d/monthly.png)](https://packagist.org/packages/cache/encrypted-cache)
+[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE)
+
+This repository has traits and interfaces that makes a PSR-6 cache implementation encrypted. Using tags allow you
+to tag related items, and then clear the cached data for that tag only. It is a part of the PHP Cache organisation. To read about
+features like tagging and hierarchy support please read the shared documentation at [www.php-cache.com](http://www.php-cache.com).
+
+*Note: Performance will be best with a driver such as memcached or redis, which automatically purges stale records.*
```
Comment (Nyholm) [inline src/Encrypted/README.md]: Should be removed

File: src/Encrypted/README.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+[![Latest Stable Version](https://poser.pugx.org/cache/encrypted-cache/v/stable)](https://packagist.org/packages/cache/encrypted-cache)
+[![codecov.io](https://codecov.io/github/php-cache/encrypted-cache/coverage.svg?branch=master)](https://codecov.io/github/php-cache/encrypted-cache?branch=master)
+[![Total Downloads](https://poser.pugx.org/cache/encrypted-cache/downloads)](https://packagist.org/packages/cache/encrypted-cache)
+[![Monthly Downloads](https://poser.pugx.org/cache/encrypted-cache/d/monthly.png)](https://packagist.org/packages/cache/encrypted-cache)
+[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE)
+
+This repository has traits and interfaces that makes a PSR-6 cache implementation encrypted. Using tags allow you
+to tag related items, and then clear the cached data for that tag only. It is a part of the PHP Cache organisation. To read about
+features like tagging and hierarchy support please read the shared documentation at [www.php-cache.com](http://www.php-cache.com).
```
Comment (Nyholm) [inline src/Encrypted/README.md]: These rows should be updated/removed.

**Comment by prisis** [comment]
@Nyholm any idea why i cant get testSaveDeferredWhenChangingValues to pass?

**Comment by Nyholm** [comment]
Ill see.

**Comment by Nyholm** [comment]
The `testSaveDeferredWhenChangingValues` test makes sure that the raw value has not changed on the deferred queue before we do a save. 

Since you encrypt the value you break the test. It is the test that is flawed though. I'll fix it.

**Comment by prisis** [comment]
Maybe i can find a workaround

File: src/Encrypted/EncryptedCachePool.php [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+/*
+ * This file is part of php-cache organization.
+ *
+ * (c) 2015-2016 Aaron Scherer <aequasi@gmail.com>, Tobias Nyholm <tobias.nyholm@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Cache\Encryption;
```
Comment (prisis) [inline src/Encrypted/EncryptedCachePool.php]: done

File: composer.json [hunk 6]
```diff
@@ -29,6 +29,7 @@
         "symfony/options-resolver": "^2.7 || ^3.0"
     },
     "require-dev": {
+        "defuse/php-encryption":   "^2.0",
```
Comment (prisis) [inline composer.json]: Is the best i know, you know a other one?

Comment (Nyholm) [inline composer.json]: I have no idea. =)

I trust that you have done some research to find a popular and robust package. And also one that you like.

Comment (prisis) [inline composer.json]: If you like you can read https://paragonie.com/blog/2016/05/defuse-security-s-php-encryption-library-version-2-0-0-released and if @paragonie-scott recommend it, so it musst be really secure :D

**Comment by prisis** [comment]
@Nyholm should i ignore this https://scrutinizer-ci.com/g/php-cache/cache/inspections/a5e25a31-409b-4228-876e-f715b2b126b5/issues/files/src/Encryption/EncryptedItemDecorator.php?status=new&orderField=path&order=asc&honorSelectedPaths=0 ?

**Comment by Nyholm** [comment]
I think we can ignore them for now but this sure is an issue.. let's fix that before making our first tag.

File: src/Encryption/EncryptedItemDecorator.php [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+/**
+ * Encrypt and Decrypt all the stored items.
+ *
+ * @author Daniel Bannert <d.bannert@anolilab.de>
+ */
+class EncryptedItemDecorator implements CacheItemInterface, HasExpirationDateInterface, TaggableItemInterface
+{
+    /**
+     * @type CacheItemInterface
```
Comment (aequasi) [inline src/Encryption/EncryptedItemDecorator.php]: IIRC these should all be `var`. I think `type` was removed from that spec

File: src/Encryption/composer.json [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+            "homepage": "https://github.com/nyholm"
+        },
+        {
+            "name":     "Daniel Bannert",
+            "email":    "d.bannert@anolilab.de",
+            "homepage": "https://github.com/prisis"
+        }
+    ],
+    "require":     {
+        "php":                   "^5.5 || ^7.0",
```
Comment (aequasi) [inline src/Encryption/composer.json]: i think `^5.5|^7.0` is the "right" way to do this now?

Also, should we continue supporting 5.5? @Nyholm

File: src/Encryption/composer.json [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        "defuse/php-encryption": "^2.0",
+        "psr/cache":             "^1.0"
+    },
+    "require-dev": {
+        "phpunit/phpunit":         "^4.0 || ^5.1",
```
Comment (aequasi) [inline src/Encryption/composer.json]: same here

File: src/Encryption/phpunit.xml.dist [hunk 10]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<phpunit backupGlobals="false"
+  backupStaticAttributes="false"
```
Comment (aequasi) [inline src/Encryption/phpunit.xml.dist]: formatting is strange here

**Comment by aequasi** [comment]
Really like this!

File: src/Encryption/composer.json (hunk 8 above)
Comment (Nyholm) [inline src/Encryption/composer.json]: Lets support 5.5 if we have no reason to drop it. 

The current syntax (at least the one in the docs) are dubble pipe.

File: src/Encryption/composer.json (hunk 9 above)
Comment (prisis) [inline src/Encryption/composer.json]: it should be double pipes https://getcomposer.org/doc/articles/versions.md

File: src/Encryption/EncryptedItemDecorator.php (hunk 7 above)
Comment (prisis) [inline src/Encryption/EncryptedItemDecorator.php]: we can fix it with styleci we use it in all packages @type

File: src/Encryption/phpunit.xml.dist (hunk 10 above)
Comment (prisis) [inline src/Encryption/phpunit.xml.dist]: we use the same format in all phpunits

**Comment by prisis** [comment]
So the last thing to do is add documentation

**Comment by prisis** [comment]
test are all green, only need to add docs for it

**Pull request merged**